src/main/java/ch/njol/skript/lang/Variable.java

```diff
@@ -294,14 +295,20 @@ public <R> Variable<R> getConvertedExpression(Class<R>... to) {
 	 * Gets the value of this variable as stored in the variables map.
 	 */
 	@Nullable
-	public Object getRaw(Event e) {
-		String n = name.toString(e);
-		if (n.endsWith(Variable.SEPARATOR + "*") != list) // prevents e.g. {%expr%} where "%expr%" ends with "::*" from returning a Map
+	public Object getRaw(Event event) {
+		String name = this.name.toString(event);
+		if (name.endsWith(Variable.SEPARATOR + "*") != list) // prevents e.g. {%expr%} where "%expr%" ends with "::*" from returning a Map
 			return null;
-		Object val = !list ? convertIfOldPlayer(n, e, Variables.getVariable(n, e, local)) : Variables.getVariable(n, e, local);
-		if (val == null)
-			return Variables.getVariable((local ? LOCAL_VARIABLE_TOKEN : "") + name.getDefaultVariableName(), e, false);
-		return val;
+		Object value = !list ? convertIfOldPlayer(name, event, Variables.getVariable(name, event, local)) : Variables.getVariable(name, event, local);
+		// Check for default variables.
+		if (value == null) {
+			for (String typeHint : this.name.getDefaultVariableNames(event)) {
```

Comment:
My only solution would be to cache the last known value from the getReturnType(event) inside the Variable object and clear it when it's updated. At least this way it would know it's type guaranteed. Do you have any alternative solution?

Reply:
Alright, I took over the TypeHints class since it was never used and was a good implementation. I created a logger handler type system that collects the runtime return type and variable names expressions' return types.